File: Assignment 1/RecommendFriend.java
package recommendfriend;

//#####################################################################
// Name: Waghe Shubham Yatindra
// Roll No.: 13MF3IM17
// Hadoop M/R Java program to find friend recommendations
// hadoop jar <jar file> RecommendFriend input/FriendData.txt output 
//#####################################################################

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.io.Writable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.*;

/**
 *
 * @author shubham
 */
public class RecommendFriend {

    /**
     * @param args the command line arguments
     */
    static public class FriendCountWritable implements Writable {
        public Long user;
        public Long mutualFriend;

        public FriendCountWritable(Long user, Long mutualFriend) {
            this.user = user;
            this.mutualFriend = mutualFriend;
        }

        public FriendCountWritable() {
            this(-1L, -1L);
        }

        @Override
        public void write(DataOutput out) throws IOException {
            out.writeLong(user);
            out.writeLong(mutualFriend);
        }

        @Override
        public void readFields(DataInput in) throws IOException {
            user = in.readLong();
            mutualFriend = in.readLong();
        }

        @Override
        public String toString() {
            return " toUser: "
                    + Long.toString(user) + " mutualFriend: " + Long.toString(mutualFriend);
        }
    }

    public static class Map extends Mapper<LongWritable, Text, LongWritable, FriendCountWritable> {
        private Text word = new Text();

        @Override
        public void map(LongWritable key, Text value, Context context) throws IOException, InterruptedException {
            String line[] = value.toString().split("\t");
            Long fromUser = Long.parseLong(line[0]);
            List<Long> toUsers = new ArrayList<Long>();

            if (line.length == 2) {
                StringTokenizer tokenizer = new StringTokenizer(line[1], ",");
                while (tokenizer.hasMoreTokens()) {
                    Long toUser = Long.parseLong(tokenizer.nextToken());
                    toUsers.add(toUser);
                    context.write(new LongWritable(fromUser), new FriendCountWritable(toUser, -1L));
                }

                for (int i = 0; i < toUsers.size(); i++) {
                    for (int j = i + 1; j < toUsers.size(); j++) {
                        context.write(new LongWritable(toUsers.get(i)), new FriendCountWritable((toUsers.get(j)), fromUser));
                        context.write(new LongWritable(toUsers.get(j)), new FriendCountWritable((toUsers.get(i)), fromUser));
                    }
                }
            }
        }
    }

    public static class Reduce extends Reducer<LongWritable, FriendCountWritable, LongWritable, Text> {
        @Override
        public void reduce(LongWritable key, Iterable<FriendCountWritable> values, Context context)
                throws IOException, InterruptedException {

            // key is the recommended friend, and value is the list of mutual friends
            final java.util.Map<Long, List<Long>> mutualFriends = new HashMap<Long, List<Long>>();

            for (FriendCountWritable val : values) {
                final Boolean isAlreadyFriend = (val.mutualFriend == -1);
                final Long toUser = val.user;
                final Long mutualFriend = val.mutualFriend;

                if (mutualFriends.containsKey(toUser)) {
                    if (isAlreadyFriend) {
                        mutualFriends.put(toUser, null);
                    } else if (mutualFriends.get(toUser) != null) {
                        mutualFriends.get(toUser).add(mutualFriend);
                    }
                } else {
                    if (!isAlreadyFriend) {
                        mutualFriends.put(toUser, new ArrayList<Long>() {
                            {
                                add(mutualFriend);
                            }
                        });
                    } else {
                        mutualFriends.put(toUser, null);
                    }
                }
            }

            java.util.SortedMap<Long, List<Long>> sortedMutualFriends = new TreeMap<Long, List<Long>>(new Comparator<Long>() {
                @Override
                public int compare(Long key1, Long key2) {
                    Integer v1 = mutualFriends.get(key1).size();
                    Integer v2 = mutualFriends.get(key2).size();
                    if (v1 > v2) {
                        return -1;
                    } else if (v1.equals(v2) && key1 < key2) {
                        return -1;
                    } else {
                        return 1;
                    }
                }
            });

            for (java.util.Map.Entry<Long, List<Long>> entry : mutualFriends.entrySet()) {
                if (entry.getValue() != null) {
                    sortedMutualFriends.put(entry.getKey(), entry.getValue());
                }
            }

            Integer i = 0;
            String output = "";
            OUTER:
            for (java.util.Map.Entry<Long, List<Long>> entry : sortedMutualFriends.entrySet()) {
                switch (i) {
                    case 0:
                        output = entry.getKey().toString();
                        break;
                    case 10:
                        break OUTER;
                    default:
                        output += "," + entry.getKey().toString();
                        break;
                }
                ++i;
            }
            context.write(key, new Text(output));
        }
    }

    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();

        Job job = new Job(conf, "RecommendFriend");
//        job.setJarByClass(RecommendFriend.class);
        job.setOutputKeyClass(LongWritable.class);
        job.setOutputValueClass(FriendCountWritable.class);

        job.setMapperClass(Map.class);
        job.setReducerClass(Reduce.class);

        job.setInputFormatClass(TextInputFormat.class);
        job.setOutputFormatClass(TextOutputFormat.class);

//        FileSystem outFs = new Path(args[1]).getFileSystem(conf);
//        outFs.delete(new Path(args[1]), true);

        FileInputFormat.addInputPath(job, new Path(args[1]));
        FileOutputFormat.setOutputPath(job, new Path(args[2]));

        job.waitForCompletion(true);
    }
    
}

File: Assignment 1/recommend_friend.py
#####################################################################
# Name: Waghe Shubham Yatindra
# Roll No.: 13MF3IM17
# Apache Spark M/R python program to find friend recommendations
# Command: spark-submit <python file> <data file>
#####################################################################
from __future__ import print_function
import sys
from pyspark.sql import SparkSession
from itertools import combinations

def map_screen_func(x):
    a = x.value.split("\t")
    return (a[0],a[1].split(','))

def emit_connections(s):
    key, friends = s[0], s[1]
    l = []
    if friends != '' and len(friends) != 0:
        for friend in friends:
            if friend != '':
                l.append((int(key), { int(friend): -1 }))
        c = combinations(friends,2)
        for f in c:
            p = list(f)
            l.append((int(p[0]), { int(p[1]): 1 }))
            l.append((int(p[1]), { int(p[0]): 1 }))
    return l

def my_reducer(p,q):
    for key in q:
        if q[key] == -1:
            p[key] = q[key]
            continue
        if key in p:
            if p[key] != -1:
                p[key] = int(p[key]) + int(q[key])
        else:
            p[key] = q[key]
    return p

def cleaner_func(x):
    values = dict(x[1])
    values = { key:val for key, val in values.items() if val != -1 }
    values = sorted(values.items(), key=lambda x: (-x[1], x[0]))
    l,i = [], 0
    for k in values:
        if i == 10:
            break
        l.append(k)
        i = i + 1
    return int(x[0]), l

def writer_func(t):
    w_string = str(t[0])
    fs = [ a[0] for a in t[1]]
    w_string += '\t'+','.join(map(str, fs))
    return w_string

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Incorrect Usage!")
        exit(-1)

    spark = SparkSession.builder.appName("RecommendFriend").getOrCreate()
    lines = spark.read.text(sys.argv[1]).rdd.map(map_screen_func)
    connections = lines.flatMap(emit_connections)
    formatter = connections.reduceByKey(lambda k,v : my_reducer(k,v)).map(cleaner_func).sortByKey()
    # formatter = connections.reduceByKey(lambda k,v : my_reducer(k,v)).map(cleaner_func).sortByKey()
    w = formatter.map(lambda t : writer_func(t))
    w.saveAsTextFile('output')
    spark.stop()
